### PR TITLE
New feature: Terminate when inactive

### LINKED ISSUE #44: Feature Request: Terminate Service bouncer after a period of inactivity.
### Overview

Service bouncer is excellent. An Ops dream tool. But... It does consume a lot of RAM. Not a single instance. But someone rolled it out across our estate and everyone keeps leaving it open. 30-40mb per instance soon adds up. Education is one thing. Automation is another :o)

### Expected Behavior

Enable the ServiceBouncer.exe to terminate itself after a period of inactivity. An hour? Lets be reasonable.

### System Info

App Version: v1.2.5.0
Operating System Version:  Windows server 2008 - 2016

### PR DESCRIPTION
**Pull request by johnwalters**
## What Changed?
New feature added: Terminate when inactive
Now the app will terminate in 60 minutes of non-user activity.

## Screenshots

serviceBouncerFeatureAppTerminate2

## Issues Closed
#44, submitted by @chrisedebo

### CODE REVIEW

**Comment by BlythMeister** [comment]
Thanks for the pull request.

I'm thinking that maybe should just make the auto kill always on?

**Comment by johnwalters** [comment]
You are welcome for the pull request. btw it is my 1st one, ever anywhere, so I am quite open to feedback. 

Sure, I considered auto kill always on, but didn’t want to disrupt. I could default to 60, and then I’d work in a way to handle the never option on the options. 
Or we could eliminate the option and hardcode to x minutes. 
Your call.

**Comment by BlythMeister** [comment]
I think hard code (at least for now) 🙂

**Comment by johnwalters** [comment]
k I'm on it. 
like a hobo on a hotdog.

**Comment by johnwalters** [comment]
## What Changed?
New feature added: Terminate when inactive
Now the app will terminate in 60 minutes of non-user activity.

## Screenshots

serviceBouncerFeatureAppTerminate2

## Issues Closed
#44, submitted by @chrisedebo

**Comment by augustoproiete** [comment]
This is nice. Thanks @johnwalters !

A couple of thoughts / possible improvements either for this PR or for a future PR:

---

**1.** The approach used here to determine inactivity is by measuring active interaction with the ServiceBouncer app specifically, and not inactivity of the user in the server. So if the user is busy interacting with another app on the server for more than the X minutes defined in ServiceBouncer, it will close when the user is technically _not_ inactive, which is not a great experience IMO...

 Maybe we could look into ways of determining inactivity of the user via Windows' [`GetLastInputInfo`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getlastinputinfo) API?

Or perhaps by checking:

* Time elapsed since the user disconnected from the RDP session on the server
* Time elapsed since the user locked his screen
* (what else?)

---

**2.** Minor nitpick: Maybe we could use a `timespan` (e.g. `01:00:00`) instead of an `int` that represents minutes, to allow for more flexibility? Of course, this requires an adjustment on the timer tick interval

**Comment by johnwalters** [comment]
@caioproiete You're welcome. On item 2, which int are you referring to?

**Comment by augustoproiete** [comment]
@johnwalters Ah! My bad. On the PR diff it shows a [change on Program.cs](https://github.com/PaulStovell/ServiceBouncer/pull/46/files#diff-98bbc2dda4856f429bff68cfc147fb14) which has some code to parse command-line arguments, and I assumed you were getting the X minutes of inactivity as a command-line argument (with a default of 60 minutes) - i.e. `ServiceBouncer.exe 90`.

I just re-read the diff and I saw no new arguments are expected and it's hard-coded to 60 minutes, so ignore item 2.

**Comment by johnwalters** [comment]
OK. re: item #1, I am open to implementing what's decided on, but I feel like more discussion is needed to nail down requirements, and I don't have anything to add to that discussion.

**Comment by BlythMeister** [comment]
I think we could do 30 minutes after session switch shows machine is locked.
So if you lock a machine for 30 minutes, the app will terminate.
But if you unlock after 25, then the timer starts again (even if you don't use Service Bouncer.

**Comment by augustoproiete** [comment]
@BlythMeister +1

We [already hook into SessionSwitch](https://github.com/PaulStovell/ServiceBouncer/blob/master/src/ServiceBouncer/MainForm.cs#L54-L64) to determine if the machine is locked or not. Seems like the right place to start/stop a timer that closes the app after 30 minutes.

**Comment by johnwalters** [comment]
I'll take a stab at implementing it. My understanding of the requirements: If the connected machine is locked and has been locked for 30 or more minutes, the app will terminate.

**Comment by johnwalters** [comment]
After having a look, I think I got that wrong. Let me try again with my understanding of the requirements: If the machine on which the application is running, is locked and has been locked for 30 or more minutes, the app will terminate.
Is that correct?

**Comment by BlythMeister** [comment]
Yes that is correct 🙂

**Comment by augustoproiete** [comment]
> If the machine on which the application is running, is locked and has been locked for 30 or more minutes, the app will terminate

@johnwalters That's correct, but it's not just if the session is locked, but also if the user has disconnected from the remote session. Either of these would be valid.

At first glance it seems you'd have to add some code to the [`SessionSwitch` event handler](https://github.com/PaulStovell/ServiceBouncer/blob/master/src/ServiceBouncer/MainForm.cs#L54) to check the `Reason` why the session switch event was raised.

* If the reason is `SessionLock` or `RemoteDisconnect`, we consider the user is **inactive**, so we start counting the time from that moment onwards. And 30 minutes from that initial time, the app closes itself.

* If the reason is `SessionUnlock` or `RemoteConnect`, the user is back so we consider the user **active**, and stop counting the time, and the app won't close itself anymore. The next time the user locks the session or disconnects from the remote session, the above will happen again and everything starts over...

ps: When receiving a `RemoteDisconnect`, we should check if we already were counting the time due to a `SessionLock` that happened moments before, so that we don't reset the counter and lose the time in between these events.

**Comment by BlythMeister** [comment]
If you have an inactive stopwatch.
When setting isactive to true, stop and reset, when setting to false start it.
When that stopwatch hits 30 minutes, kill the app.

**Comment by johnwalters** [comment]
the class variable isActive is used not only when the session state changes, but also when the form is activated or deactivated, meaning when the user clicks on another app on the same box as this app or back onto this app. So I don't think that is what we want, is it? I'll try again:
1)  If the machine on which the application is running, is (locked or remotely disconnected) and has been (locked or remotely disconnected) for 30 or more minutes, the app will terminate.
That makes sense to me. Is that correct?

**Comment by augustoproiete** [comment]
> If the machine on which the application is running, is (locked or remotely disconnected) and has been (locked or remotely disconnected) for 30 or more minutes, the app will terminate.

Yup. That's it. The requirements remain the same. @BlythMeister was just suggesting a possible approach to implement that... Which from your observation, doesn't look like it's a fit.

**Comment by johnwalters** [comment]
## What Changed?
New feature added: Terminate when inactive
Now the app will terminate if machine on which the app is running is either locked or remotely disconnected, and 30 or more minutes have passed.

## Screenshots
(no UI change)

## Issues Closed
#44, submitted by @chrisedebo

File: README.md [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
-
-**Download Zip**: [ServiceBouncer.zip](https://github.com/PaulStovell/ServiceBouncer/releases)
-
-**Download Octopus Nuget Package**: [ServiceBouncer.nupkg](https://github.com/PaulStovell/ServiceBouncer/releases)
-
-The tool shows a data grid for the services, and their status. Since it's a grid, you can easily select multiple services. You can then start or stop them. You can also filter the services shown, so that you don't accidentally stop the wrong one. There's even a button to delete services you no longer need. 
-
-![ServiceBouncer main window - stop, start, delete, filter & manage services](https://raw.githubusercontent.com/PaulStovell/ServiceBouncer/master/.images/MainWindow.png)
-
+As a developer building Windows Services, the workflow of constantly stopping and starting services can be a bit annoying, especially if your solution consists of multiple services. **ServiceBouncer** is a simple tool that helps to streamline this workflow.
```
Comment (BlythMeister) [inline README.md]: I can't see the differences in this file

File: src/ServiceBouncer/MainForm.cs [hunk 2]
```diff
@@ -19,6 +19,10 @@ public partial class MainForm : Form
         private bool isActive;
         private string machineHostname;
         private int backgroundRefreshSeconds;
+        private DateTime? machineInactivatedTime;
+        private int inactivityMinutesUntilAppTermination = 30;
+        private System.Windows.Forms.Timer appTerminationTimer = new System.Windows.Forms.Timer();
```
Comment (BlythMeister) [inline src/ServiceBouncer/MainForm.cs]: Could you use System.Threading.Timer instead

File: src/ServiceBouncer/MainForm.cs [hunk 3]
```diff
@@ -28,7 +32,10 @@ public MainForm(string machine)
             machineHostname = machine;
             toolStripConnectToTextBox.Text = machineHostname;
             services = new List<ServiceViewModel>();
-            Microsoft.Win32.SystemEvents.SessionSwitch += SessionSwitch;
+            Microsoft.Win32.SystemEvents.SessionSwitch += SessionSwitch;
+            appTerminationTimer.Tick += new EventHandler(TerminateIfInactive);
+            appTerminationTimer.Interval = 60000;
```
Comment (BlythMeister) [inline src/ServiceBouncer/MainForm.cs]: When setting an interval in milliseconds, use Timespan.FromMinutes(1).TotalMilliseconds.
Or FromSeconds etc as this is clearer when reading what the value is without needing to do maths.

File: src/ServiceBouncer/MainForm.cs [hunk 4]
```diff
@@ -387,17 +399,30 @@ private void SetTitle()
         }
 
         private void SetConnectedStatusBar()
-        {
+        {
             if (isActive)
             {
-                var backgroundRefreshTimeText = backgroundRefreshSeconds == 1 ? "1 second" : $"{backgroundRefreshSeconds} seconds";
+                var backgroundRefreshTimeText = backgroundRefreshSeconds == 1 ? "1 second" : $"{backgroundRefreshSeconds} seconds";
```
Comment (BlythMeister) [inline src/ServiceBouncer/MainForm.cs]: Can't see the differences in this or the above line

File: src/ServiceBouncer/MainForm.cs [hunk 5]
```diff
@@ -19,6 +19,10 @@ public partial class MainForm : Form
         private bool isActive;
         private string machineHostname;
         private int backgroundRefreshSeconds;
+        private DateTime? machineInactivatedTime;
+        private int inactivityMinutesUntilAppTermination = 30;
```
Comment (BlythMeister) [inline src/ServiceBouncer/MainForm.cs]: If you make this a timespan it make the checking easier.  You can say is this timespan bigger that other timespan.

File: src/ServiceBouncer/MainForm.cs [hunk 6]
```diff
@@ -387,17 +399,30 @@ private void SetTitle()
...
             }
             else
             {
                 toolStripStatusLabel.Text = $@"Connected to {machineHostname}. - Background refresh disabled";
             }
         }
+
+        private void TerminateIfInactive(Object obj, EventArgs args)
+        {
+            // Check to see if machine has been in an inactive state for inactivityMinutesUntilAppTermination
```
Comment (BlythMeister) [inline src/ServiceBouncer/MainForm.cs]: You don't really need this comment, the function is clear enough without it.

File: src/ServiceBouncer/MainForm.cs [hunk 7]
```diff
@@ -387,17 +399,30 @@ private void SetTitle()
...
+            // Inactive state caused by machine (where app is running) being locked or remotely disconnected
+            // If so, terminate this app
+            if (!machineInactivatedTime.HasValue) return;
+            var minutesSinceInactivated = (DateTime.Now - machineInactivatedTime.Value).TotalMinutes;
+            if (minutesSinceInactivated > inactivityMinutesUntilAppTermination)
```
Comment (BlythMeister) [inline src/ServiceBouncer/MainForm.cs]: This if becomes different if using timespans

File: src/ServiceBouncer/MainForm.cs [hunk 8]
```diff
@@ -387,17 +399,30 @@ private void SetTitle()
...
+        private void TerminateIfInactive(Object obj, EventArgs args)
+        {
+            // Check to see if machine has been in an inactive state for inactivityMinutesUntilAppTermination
+            // Inactive state caused by machine (where app is running) being locked or remotely disconnected
+            // If so, terminate this app
+            if (!machineInactivatedTime.HasValue) return;
+            var minutesSinceInactivated = (DateTime.Now - machineInactivatedTime.Value).TotalMinutes;
+            if (minutesSinceInactivated > inactivityMinutesUntilAppTermination) 
+            {
+                Process.GetCurrentProcess().Kill();
```
Comment (BlythMeister) [inline src/ServiceBouncer/MainForm.cs]: You should use Application.Exit as it's a nicer kill 🤣

File: src/ServiceBouncer/Program.cs [hunk 9]
```diff
@@ -1,32 +1,32 @@
...
-                    Application.Run(new MainForm(Environment.MachineName));
-                });
-        }
-    }
-}
+﻿using CommandLine;
+using System;
+using System.Windows.Forms;
+
+namespace ServiceBouncer
```
Comment (BlythMeister) [inline src/ServiceBouncer/Program.cs]: Can't see any difference here...

File: src/ServiceBouncer/ServiceBouncer.csproj [hunk 10]
```diff
@@ -1,212 +1,212 @@
...
-        <Reference Include="CredentialManagement">
-          <HintPath>..\..\packages\CredentialManagement\lib\net35\CredentialManagement.dll</HintPath>
-          <Private>True</Private>
-          <Paket>True</Paket>
-        </Reference>
-      </ItemGroup>
-    </When>
-  </Choose>
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (BlythMeister) [inline src/ServiceBouncer/ServiceBouncer.csproj]: Another file which doesn't need to be here as the same.

**Review by BlythMeister — CHANGES_REQUESTED** [review]
A few changes and a few files that shouldn't be in the or at all as they are the same

File: src/ServiceBouncer/MainForm.cs [hunk 11]
```diff
@@ -55,11 +62,16 @@ private void SessionSwitch(object sender, Microsoft.Win32.SessionSwitchEventArgs
         {
             if (e.Reason == Microsoft.Win32.SessionSwitchReason.SessionLock || e.Reason == Microsoft.Win32.SessionSwitchReason.RemoteDisconnect)
             {
-                isActive = false;
+                isActive = false;
+                if (!machineInactivatedTime.HasValue)
+                {
+                    machineInactivatedTime = DateTime.Now;
```
Comment (augustoproiete) [inline src/ServiceBouncer/MainForm.cs]: You could schedule the timer here, for `machineInactivatedTime + 30 minutes`, that way you don't have to run the timer every single minute to check the inactive time.

File: src/ServiceBouncer/MainForm.cs [hunk 12]
```diff
@@ -55,11 +62,16 @@ private void SessionSwitch(object sender, Microsoft.Win32.SessionSwitchEventArgs
...
+                if (!machineInactivatedTime.HasValue)
+                {
+                    machineInactivatedTime = DateTime.Now;
+                }
             }
             else if (e.Reason == Microsoft.Win32.SessionSwitchReason.SessionUnlock || e.Reason == Microsoft.Win32.SessionSwitchReason.RemoteConnect)
             {
-                isActive = true;
+                isActive = true;
+                machineInactivatedTime = null;
```
Comment (augustoproiete) [inline src/ServiceBouncer/MainForm.cs]: You could disable the timer here, and it will not run until a new session switch happens again

File: src/ServiceBouncer/MainForm.cs (hunk 11 above)
Comment (BlythMeister) [reply src/ServiceBouncer/MainForm.cs]: Yeah, that's a nice approach

**Comment by johnwalters** [comment]
ok please have a look.

**Review by BlythMeister — APPROVED** [review]
Very nice thank you.

**Comment by johnwalters** [comment]
Thanks for the help guys.

**Comment by BlythMeister** [comment]
No worries, thanks for the changes.
I'll try and get a release out today

**Comment by chrisedebo** [comment]
Hooray! 

Thanks for adding my feature request.

Good work chaps :)
